openedx/core/djangoapps/plugins/README.rst

```diff
@@ -189,14 +189,27 @@ class::
             # Configuration setting for Plugin Contexts for this app.
             PluginContexts.CONFIG: {
 
-                # Configure the Plugin Signals for each Project Type, as needed.
+                # Configure the Plugin Contexts for each Project Type, as needed.
                 ProjectType.LMS: {
 
                     # Key is the view that the app wishes to add context to and the value
                     # is the function within the app that will return additional context
                     # when called with the original context
                     u'course_dashboard': u'my_app.context_api.get_dashboard_context'
                 }
+            },
+
+            # Configuration settings for Plugin Slots for this app.
+            PluginSlots.CONFIG: {
+
+                # Configure Plugin Slots for each Project Type, as needed.
+                ProjectType.LMS: {
+
+                    # Key is the slot in the current Project Type that the app wants to inject
+                    # content into and the value is the function within the app that will return
+                    # rendered content. The function will be passed minimal context relevant to it.
+                    u'head-extra': u'my_app.slots_api.get_slot_content'
```

Comment:
Random thought as I look through - it may be that we'll want to namespace these by the view name to avoid naming collisions/ambiguity.  Maybe something like:

```
PluginSlots.CONFIG: {
    ProjectType.LMS: {
        u'course_dashboard': { # identifier for the view 
            u'head-extra': u'my_app.slots_api.get_slot_content'
        }
    }
}
```

Reply:
@davidjoy I think that might make things harder for many usecases. For instance, if you want to inject something to every view in the header, then this approach will make that harder. On the other hand, the plugin itself can always filter based on the URL if we give it enough info. 

Alternatively we can have an all-or-nothing approach and also support something like:

```python
PluginSlots.CONFIG: {
    ProjectType.LMS: {
        # ...
        u'*': { # all views
            u'head-extra': u'my_app.slots_api.get_slot_content'
        },
    }
}
```

However, the bigger issue is, how do you filter this based on the view name or id. We can pass in the URL, but I'm not sure how to automatically pass in the view name. We'd need to explicitly add this to each view, potentially by adding something like 'view-name' to the context.

Also, some slots can be in templates that only exist in certain views. In that case we can have automatic namespacing, as the slot names can be picked appropriately. For instance 'instructor-dashboard-download-end' or something like that.

Reply:
I think the danger in having a wildcard for the view is that slot _names_ may be shared across views, but the intent/location of those slot names may not be the same.  If you use a wildcard to apply something to "head-extra" across the entire site, it'll be hard to guarantee that the identifier "head-extra" means the same thing to all pages.  Some page with a table on it, for instance, might have a slot for adding additional header rows. ¯\_(ツ)_/¯ (made up example, obviously)

If we consider the LMS header, studio header, etc., to be their own namespaces, then a plugin applied to the 'head-extra' slot in the LMS header will then apply across all pages that use that template.  

Maybe thinking about the namespaces as the view names, specifically, is too constrained.  I think they should be 'domain' categories and the convention is that their string identifier corresponds to a domain.  Slots in the instructor dashboard would have a namespace of 'instructor-dashboard' regardless of the specific view within that.

This maps onto how we're thinking of plugins/slots for the micro-frontends as well, though those will be setup via a different mechanism.  The MFE itself is the namespace and has an identifier like "profile", "learning", "account", "payment", and then the slot names need to be unique in that bucket.

So a slot would be added to the template with a name and possibly a namespace; we may be able to find a way to declare the namespace in the app/view and make it available via context so it doesn't have to be hard-coded on each slot in the templates.  That's just a nicety, though.  

When the template is rendering, it evaluates the slot and looks it up in the plugin slots config by its name and namespace (since it's just a dict), and that tells it what plugin(s) are registered for that slot.  Should be pretty efficient since it's just a dict lookup of a few strings; not view names or URLs or anything difficult to resolve programmatically.

Let me know if I'm not making sense!

Reply:
I think that makes sense. 

I do think there can be a use case for adding some content globally, for instance tracking codes in the page head and footer. In such a case having a global option would prevent the need to update the plugin each time a new domain is introduced. Perhaps one solution is to have just these three slots available globally and name them appropriately so there is no chance of conflict.

Here is what I'm picturing then:
- Have a `lms` and `studio` namespaces that apply to all pages in those platforms. 
- Rename the slots to something like:
  - `head-extra` -> `page-header-extra-content`
  - `body-initial` -> `page-body-initial-content`
  - `body-extra` -> `page-body-extra-content`
- These slots can be part of the `lms` and `studio` namespaces, and in addition, they can also apply to the specific domains as you described them. 
- For instance the `page-header-extra-content` could exist for both the `lms` and for `instructor-dashboard`, which could optionally be `lms:instructor-dashboard` to further segregate it from potential other uses.
- Each `lms` slot can automatically apply to `lms` and in addition the specific app in the lms.
- The slot name can be defined in the template while the namespace/domain can come from the app name and 'lms' / 'studio' can be applied globally in the respective folders. 

Note that in the end what I was trying to accomplish was the same effect as adding a "head-extra.html" file with some content in it, but make it available as a plugin API. I would love it if that simplicity was also available while allowing for more flexible domain/namespace-based content injection.

Does that sound OK? If the global slot option still seems concerning I can proceed with just the namespaced slots.

Reply:
I think this sounds okay, yes.  

I'm thinking a bit about the idea of 'plugin inheritance', i.e., a page inheriting any plugins defined as part of the application namespace, and wondering if that can instead be solved by just having two slots:

1. One slot for plugins from a 'common' domain.  "common:head-extra"
2. One slot for plugins for the page-specific domain, right below that.  "instructor-dashboard:head-extra".

My gut says that we shouldn't allow for blanket, 'namespace' level application of plugins to slots because it invites the same sort of problems as the wildcard.  It means pages can't re-use slot names for semantically different slots, even if it makes sense within their domain.  A "navigation" slot, for instance, could mean completely different things on differnet pages (and even have different shapes/screen real-estate).  Applying a plugin to "lms:navigation" could have pretty undefined behavior, especially if we don't even know all of the apps that have defined a "navigation" plugin slot! 

i.e., I'm wondering if we treat 'namespace' level things as just another domain that has slots on a wider variety of pages, or in common templates.  Adding a new "common" slot should be rare.

Reply:
I _do_ imagine all this will evolve, of course.  I wonder if we start simple with a single-layer system for now, though.  (So I suppose I'm voting for proceeding with just the domain slots without LMS/Studio-level namespaces)

Reply:
Sure, I think it makes sense to start with the restricted set and see how things evolve and if a more global slot becomes necessary. Will adjust my PR accordingly and get back to you.

Reply:
@davidjoy Sorry for the long delay!

I have updated this PR to add support for namespaces. 
It seems there was already a discussions middleware that was injecting the view name into the request, so I took it one step forward, and added support for letting a view define this as well. 

This allows a plugin to use the current view names already while also supporting overriding this so that multiple views can have the same namespace.